---
title: "coala_sims"
output: html_document
date: "2023-10-17"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(snpR); library(data.table); library(coala);
# remotes::install_github("hemstrow/snpR")
```

# Define parameters

```{r}
N0 <- 10000 # starting population size (at present)
mu <- 1e-8 # per-base mutation rate
len <- 10000000 # chromosome length
r <- 1 # average number of recombination events per chr
g <- 1 # years per gen
t1 <- 500 # time of ingroup split
t2 <- 1000 # time of outgroup split
tb <- 50 # time of bottleneck/growth for size change models -- for the slection model this is 50
m <- .1/N0 # number of migrants per gen, near isolation (Fst ~ .1); 10/N0, 50/N0, 100/N0 for low geneflow, mod geneflow, high geneflow
change_size <- 10 # proportion change, .05 for bottle, 10 for expansion
samples <- 50 # sample size
length_growth <- 5 # number of generations where population size is ramping up or down
nloci <- 10
do_size_change <- FALSE

output_file_name <- "../data/simulated/soft_sweeps" # no file extension
```

# Define models
## Basic model
Make a three-population model where the outgroup splits from the two ingroup populations *t2* generations in the past. The two ingroup populations then split *t1* generations before the present (1 gen. = 1 year). Each population (unless with size change) has *N0* individuals at the present time. The genome is modeled as *nloci* chromosomes of length *len* bp, with a per-base mutation rate *mu* and recombination rate *r* (average number of recombination events per chromosome). A "moderate" level of gene flow is modeled in the basic model, with 50 migrants per generation, representing 50/*N0* proportion of the population being replaced by migrants. We sample *samples* individuals from each population at the end of the simulation. To add selection and demographic changes, we simply build on this basic model.
```{r}
model_basic <- coal_model(sample_size = c(samples, samples, samples), loci_number = nloci, ploidy = 2, loci_length = len) +
  feat_mutation(rate = par_expr(4*N0*mu*len)) + # 4*N*mu, where mu is mut. rate per locus
  feat_recombination(rate = par_expr(4*N0*r)) + # 4*N*r, where r is the probability that a recombination event within the locus occurs in one generation.
  feat_pop_merge(par_expr(t1/(4*N0*g)), 3, 2) + #pop. 3 merges into pop. 2, 0.5 coal. units in the past
  feat_pop_merge(par_expr(t2/(4*N0*g)), 2, 1) + #pops. 2+3 merge into pop. 1, 0.3 coal. units before 2+3 divergence
  sumstat_seg_sites() + # generate segregating snps
  par_named("N0") +
  par_named("mu") +
  par_named("r") +
  par_named("g") +
  par_named("t1") +
  par_named("t2")
  


if(!do_size_change){
  model_basic <- model_basic + 
    feat_migration(rate = par_expr(4 * N0 * m), pop_from = 2, pop_to = 3) + 
    feat_migration(rate = par_expr(4 * N0 * m), pop_from = 3, pop_to = 2) + 
    par_named("m")
}

```

## bottleneck/expansion
To the basic model, we add a exponential rate change (either population expansion or contraction) that begins *tb / (4 x N0 x g)* coalescent units ago, or *tb* generations ago, in population 3 and contiunes for *length_growth/(4 x N0 x g)* coalescent units (or *length_growth* generations). A bottleneck is represented as the population shrinking to a population size that is 5% of the original. A population expansion represents a growth to a final size of 10x the ancestral population size. The population size change occurs over 5 generations, and then stays at that size until the present. We set migration to very, very low (.1/g) to reduce the impact of geneflow on the expansion/bottleneck signatures.
```{r}
# alpha (rate values) are from the equation Nt = N0e^(-a*t)
# rearranged to alpha = -ln(N1/N0)/t
model_size_change <- model_basic +
  feat_size_change(new_size = par_named("b_size"), 3, time = 0) +
  feat_growth(rate = 0, time = 0, population = 3) + 
  feat_growth(rate = par_expr(-(log(1/b_size))/(length_growth/(4*N0*g))), 
              time = par_expr((tb - length_growth)/(4*N0*g)),
              population = 3) +
  feat_growth(rate = 0, time = par_expr(tb/(4*N0*g)), population = 3) +
  # feat_size_change(new_size = 1, 3, time = par_expr(tb/(4*N0*g))) +
  par_named("tb") +
  par_named("length_growth")
```

# run model

We run the the basic/size_change/selection models according to the settings above. Note that we need to use msms for the selection models and need to set the paths for both msms and java.
```{r}
if(!do_size_change){
  sim <- simulate(model_basic, seed = 123, pars = c(N0 = N0, len = len,
                                                    r = r, g = g, t1 = t1, t2 = t2, 
                                                    tb = tb, mu = mu, m = m))
}

if(do_size_change){
  sim <- simulate(model_size_change, seed = 123, pars = c(N0 = N0, len = len,
                                                          r = r, g = g, t1 = t1, t2 = t2, 
                                                          tb = tb, mu = mu, 
                                                          b_size = change_size,
                                                          length_growth = length_growth))
}


```

# parse
```{r}

genos <- sim$seg_sites
meta <- vector("list", length(genos))
for(i in 1:length(genos)){
  meta[[i]] <- data.table(chr = i, position = floor(as.numeric(genos[[i]]$position)*len))
  genos[[i]] <- genos[[i]]$snps[seq(1, nrow(genos[[i]]), 2),] +
    genos[[i]]$snps[seq(2, nrow(genos[[i]]), 2),]
  genos[[i]] <- t(genos[[i]])
  genos[[i]] <- as.data.table(genos[[i]])
}

meta <- rbindlist(meta)
genos <- rbindlist(genos)

# saveRDS(list(genos = genos, meta = meta), paste0(output_file_name, ".RDS"))

# send to vcf
srd <- import.snpR.data(genos, sample.meta = data.frame(pop = rep(c("A", "B", "C"), each = 50)),
                        snp.meta = meta)

# format_snps(srd, output = "vcf", outfile = paste0(output_file_name, ".vcf"))

# 
plot_clusters(srd, "pop")$plot
srd <- calc_tajimas_d(srd, "pop", global = TRUE)
get.snpR.stats(srd, "pop", "tajimas_d")$weighted.means

# plot_sfs(srd, facet = "pop", pops = c("C", "B"), projection = c(90, 90), log = TRUE)
srd <- calc_pairwise_fst(srd, "pop")
get.snpR.stats(srd, "pop", "fst")$weighted.means
# 
# # selection detection
# # srd <- calc_pi(srd, "pop")
# # srd <- calc_fis(srd, "pop")
plot_manhattan(srd, "fst", facets = "pop")


# srd <- srd[pop = c("B", "C")]
# keep_samps <- sort(c(sample(1:50, 30, FALSE), sample(51:100, 30, FALSE)))
# srd <- srd[,keep_samps]
# srd <- calc_maf(srd)
# srd <- filter_snps(srd, non_poly = TRUE)
# saveRDS(srd, "../data/simulated/neutral.RDS")

```
